app/cdap/components/PipelineDiff/util/diff.ts

```diff
@@ -0,0 +1,185 @@
+ */
+function preprocessPipeline(pipeline: IPipelineConfig) {
+  const stageMap: IStageMap = {};
+  const connectionMap: IConnectionMap = {};
+  const stageNameToStage = {};
+
+  // Arrays are compared in json-diff in an ordered fashion. To overcome this issue, convert
+  // stages and connections array into objects with each object/connection having their unique key
+
+  for (const stage of pipeline.stages) {
+    stageMap[getStageDiffKey(stage)] = stage;
+    stageNameToStage[stage.name] = stage;
+  }
+
+  // since stage names are not enough to generate a unique key, and connections only use the stage
+  // name. using this function that allows for generation of unique connection keys.
+  const getStageDiffKeyFromStageName = (name: string) => getStageDiffKey(stageNameToStage[name]);
+
+  for (const connection of pipeline.connections) {
+    const connectionMapKey = getConnectionDiffKey(connection, getStageDiffKeyFromStageName);
+    connectionMap[connectionMapKey] = connection;
+  }
+  return { stageMap, connectionMap };
+}
+
+/**
+ * Types the diff function from json-diff for our use-case.
+ * @param obj1 obj1
+ * @param obj2 obj2
+ * @returns difference between obj1 and obj2
+ */
+function jsonObjectDiff<T1>(obj1: T1, obj2: T1) {
+  return diff(obj1, obj2) as { [name: string]: DeepPartial<T1> };
+}
+
+/**
+ * Given two pipelines, return the difference between them
+ * @param pipeline1 first pipeline
+ * @param pipeline2 second pipeline
+ * @returns edit actions applied to pipeline1 to get pipeline2
+ */
+export function computePipelineDiff(pipeline1: IPipelineConfig, pipeline2: IPipelineConfig) {
+  const { stageMap: stageMap1, connectionMap: connectionMap1 } = preprocessPipeline(pipeline1);
+  const { stageMap: stageMap2, connectionMap: connectionMap2 } = preprocessPipeline(pipeline2);
+
+  // json-diff appends _added or _deleted if an object property has been added or deleted
+  // if a property is modified, no name changes happen
+  // no property appears in the diff object if the property has not changed
+  const stagesDiffMap = jsonObjectDiff(stageMap1, stageMap2);
+  const connectionsDiffMap = jsonObjectDiff(connectionMap1, connectionMap2);
+
+  const diffMap: IPipelineDiffMap = {
+    stages: {},
+    connections: {},
+  };
+
+  Object.keys(stagesDiffMap).forEach((stage) => {
+    let stageName = stage;
+    let diffIndicator = DiffIndicator.MODIFIED;
+
+    if (stage.match(/__added$/)) {
+      stageName = stage.replace(/__added$/, '');
+      diffIndicator = DiffIndicator.ADDED;
+    } else if (stage.match(/__deleted$/)) {
+      stageName = stage.replace(/__deleted$/, '');
+      diffIndicator = DiffIndicator.DELETED;
+    }
+    diffMap.stages[stageName] = {
+      diffIndicator,
+      stage1: stageMap1[stageName],
+      stage2: stageMap2[stageName],
+      diff: stagesDiffMap[stage],
+    };
+  });
+  Object.keys(connectionsDiffMap).forEach((connection) => {
```

Comment:
Seems like you need to handle cases when `connectionsDiffMap` is undefined

Reply:
Fixed.